flink-examples/flink-examples-cep/src/main/java/org/apache/flink/cep/examples/java/monitoring/TemperatureMonitoring.java

```diff
@@ -0,0 +1,147 @@
+ * This is achieved by defining another CEP pattern which analyzes the stream of generated
+ * temperature warnings.
+ */
+public class TemperatureMonitoring {
+
+	private static final double TEMPERATURE_THRESHOLD = 100;
+
+	public static void main(String[] args) throws Exception {
+		System.out.println("Executing temperature monitoring Java example.");
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		// Use ingestion time => TimeCharacteristic == EventTime + IngestionTimeExtractor
+		env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
+
+		// Input stream of monitoring events
+		DataStream<MonitoringEvent> inputEventStream = env.addSource(new MonitoringEventSource())
+			.assignTimestampsAndWatermarks(new IngestionTimeExtractor<>());
+
+		// Warning pattern: Two consecutive temperature events whose temperature is higher
+		// than the given threshold appearing within a time interval of 10 seconds
+		Pattern<MonitoringEvent, ?> warningPattern = Pattern
+			.<MonitoringEvent>begin("first")
+				.subtype(TemperatureEvent.class)
+				.where(new SimpleCondition<TemperatureEvent>() {
+					@Override
+					public boolean filter(TemperatureEvent event) throws Exception {
+						return event.getTemperature() > TEMPERATURE_THRESHOLD;
+					}
+				})
+			.next("second")
+				.subtype(TemperatureEvent.class)
+				.where(new SimpleCondition<TemperatureEvent>() {
+					@Override
+					public boolean filter(TemperatureEvent event) throws Exception {
+						return event.getTemperature() > TEMPERATURE_THRESHOLD;
+					}
+				})
+			.within(Time.seconds(10));
+
+		// Create a pattern stream from our warning pattern
+		PatternStream<MonitoringEvent> tempPatternStream = CEP.pattern(
+			inputEventStream.keyBy("rackID"),
+			warningPattern);
+
+		// Generate temperature warnings for each matched warning pattern
+		DataStream<TemperatureWarning> warnings = tempPatternStream.select(
+			new PatternSelectFunction<MonitoringEvent, TemperatureWarning>() {
+				@Override
+				public TemperatureWarning select(
+						Map<String, List<MonitoringEvent>> pattern) throws Exception {
+					TemperatureEvent first = (TemperatureEvent) pattern.get("first").get(0);
+					TemperatureEvent second = (TemperatureEvent) pattern.get("second").get(0);
+					return new TemperatureWarning(
+						first.getRackID(),
+						(first.getTemperature() + second.getTemperature()) / 2
+					);
+				}
+			}
+		);
+
+		// Alert pattern: Two consecutive temperature warnings
+		// appearing within a time interval of 20 seconds
+		Pattern<TemperatureWarning, ?> alertPattern = Pattern
+			.<TemperatureWarning>begin("first")
+			.next("second")
+			.within(Time.seconds(20));
+
+		// Create a pattern stream from our alert pattern
+		PatternStream<TemperatureWarning> alertPatternStream = CEP.pattern(
+			warnings.keyBy("rackID"),
+			alertPattern);
+
+		// Generate a temperature alert iff the second temperature warning's average temperature
```

Comment:
type `iff`

Reply:
Not a typo, just an abbreviation for `if and only if` :)

Reply:
lol, didn't know lol